File: backend-nestjs/src/modules/products/batch-consumption.service.ts
import { Injectable } from "@nestjs/common";
import { InjectRepository } from "@nestjs/typeorm";
import { EntityManager, Repository } from "typeorm";
import { Product } from "./entities/product.entity";
import { ProductBatch } from "./entities/product-batch.entity";
import { OrderItemBatch } from "../orders/entities/order-item-batch.entity";
import { BusinessException } from "../../common/exceptions/business.exception";

@Injectable()
export class BatchConsumptionService {
  constructor(
    @InjectRepository(ProductBatch)
    private readonly productBatchRepository: Repository<ProductBatch>,
  ) {}

  async consumeFefo(
    manager: EntityManager,
    productId: number,
    quantity: number,
  ): Promise<
    { batchId: number; quantityTaken: number; expiryDate: string | null }[]
  > {
    if (quantity <= 0) {
      return [];
    }

    const product = await manager
      .getRepository(Product)
      .createQueryBuilder("p")
      .setLock("pessimistic_write")
      .where("p.id = :productId AND p.deleted_at IS NULL", { productId })
      .getOne();

    if (!product) {
      throw new BusinessException(
        "PRODUCT_NOT_FOUND",
        404,
        `Không tìm thấy sản phẩm ID ${productId}.`,
      );
    }

    const batches = await manager
      .getRepository(ProductBatch)
      .createQueryBuilder("pb")
      .setLock("pessimistic_write")
      .where(
        "pb.product_id = :productId AND pb.deleted_at IS NULL AND pb.quantity_remaining > 0",
        {
          productId,
        },
      )
      .orderBy("pb.expiry_date", "ASC", "NULLS LAST")
      .addOrderBy("pb.id", "ASC")
      .getMany();

    const totalRemaining = batches.reduce(
      (sum, b) => sum + b.quantityRemaining,
      0,
    );
    if (totalRemaining < quantity) {
      throw new BusinessException(
        "INVENTORY_INSUFFICIENT",
        409,
        `Sản phẩm "${product.name}" không đủ tồn kho (còn ${totalRemaining}, cần ${quantity}).`,
      );
    }

    let quantityToConsume = quantity;
    const consumed: {
      batchId: number;
      quantityTaken: number;
      expiryDate: string | null;
    }[] = [];

    for (const batch of batches) {
      if (quantityToConsume <= 0) break;

      const take = Math.min(batch.quantityRemaining, quantityToConsume);
      batch.quantityRemaining -= take;
      quantityToConsume -= take;

      consumed.push({
        batchId: batch.id,
        quantityTaken: take,
        expiryDate: batch.expiryDate,
      });
      await manager.getRepository(ProductBatch).save(batch);
    }

    product.stockQuantity -= quantity;

    const earliestBatch = await manager
      .getRepository(ProductBatch)
      .createQueryBuilder("pb")
      .where(
        "pb.product_id = :productId AND pb.deleted_at IS NULL AND pb.quantity_remaining > 0",
        {
          productId,
        },
      )
      .orderBy("pb.expiry_date", "ASC", "NULLS LAST")
      .addOrderBy("pb.id", "ASC")
      .getOne();

    product.nearestExpiryDate = earliestBatch ? earliestBatch.expiryDate : null;
    await manager.getRepository(Product).save(product);

    return consumed;
  }

  async consumeSpecificBatch(
    manager: EntityManager,
    productId: number,
    batchId: number,
    quantity: number,
  ): Promise<
    { batchId: number; quantityTaken: number; expiryDate: string | null }[]
  > {
    if (quantity <= 0) {
      return [];
    }

    const product = await manager
      .getRepository(Product)
      .createQueryBuilder("p")
      .setLock("pessimistic_write")
      .where("p.id = :productId AND p.deleted_at IS NULL", { productId })
      .getOne();

    if (!product) {
      throw new BusinessException(
        "PRODUCT_NOT_FOUND",
        404,
        `Không tìm thấy sản phẩm ID ${productId}.`,
      );
    }

    const batch = await manager
      .getRepository(ProductBatch)
      .createQueryBuilder("pb")
      .setLock("pessimistic_write")
      .where(
        "pb.id = :batchId AND pb.product_id = :productId AND pb.deleted_at IS NULL",
        { batchId, productId },
      )
      .getOne();

    if (!batch) {
      throw new BusinessException(
        "BATCH_NOT_FOUND",
        404,
        `Không tìm thấy lô hàng ID ${batchId} của sản phẩm này.`,
      );
    }

    if (batch.quantityRemaining < quantity) {
      throw new BusinessException(
        "INVENTORY_INSUFFICIENT",
        409,
        `Lô hàng "${batch.batchCode}" chỉ còn tồn ${batch.quantityRemaining} (cần trừ ${quantity}).`,
      );
    }

    batch.quantityRemaining -= quantity;
    await manager.getRepository(ProductBatch).save(batch);

    product.stockQuantity -= quantity;

    const earliestBatch = await manager
      .getRepository(ProductBatch)
      .createQueryBuilder("pb")
      .where(
        "pb.product_id = :productId AND pb.deleted_at IS NULL AND pb.quantity_remaining > 0",
        { productId },
      )
      .orderBy("pb.expiry_date", "ASC", "NULLS LAST")
      .addOrderBy("pb.id", "ASC")
      .getOne();

    product.nearestExpiryDate = earliestBatch ? earliestBatch.expiryDate : null;
    await manager.getRepository(Product).save(product);

    return [
      {
        batchId: batch.id,
        quantityTaken: quantity,
        expiryDate: batch.expiryDate,
      },
    ];
  }

  async receiveBatch(
    manager: EntityManager,
    productId: number,
    quantity: number,
    expiryDate: string | null,
    unitCost: number,
    createdBy: number,
    batchCode?: string,
  ): Promise<ProductBatch> {
    const product = await manager
      .getRepository(Product)
      .createQueryBuilder("p")
      .setLock("pessimistic_write")
      .where("p.id = :productId AND p.deleted_at IS NULL", { productId })
      .getOne();

    if (!product) {
      throw new BusinessException(
        "PRODUCT_NOT_FOUND",
        404,
        `Không tìm thấy sản phẩm ID ${productId}.`,
      );
    }

    const batchRepo = manager.getRepository(ProductBatch);
    const code = batchCode && batchCode.trim() ? batchCode.trim() : null;

    const newBatch = batchRepo.create({
      productId,
      batchCode: code || "TEMP",
      quantityReceived: quantity,
      quantityRemaining: quantity,
      unitCost,
      expiryDate: expiryDate || null,
      createdBy,
    });

    const savedBatch = await batchRepo.save(newBatch);

    if (!code) {
      const now = new Date();
      const yyyy = now.getFullYear();
      const mm = String(now.getMonth() + 1).padStart(2, "0");
      const dd = String(now.getDate()).padStart(2, "0");
      savedBatch.batchCode = `LÔ-${yyyy}${mm}${dd}-${savedBatch.id}`;
      await batchRepo.save(savedBatch);
    }

    product.stockQuantity += quantity;

    const earliestBatch = await manager
      .getRepository(ProductBatch)
      .createQueryBuilder("pb")
      .where(
        "pb.product_id = :productId AND pb.deleted_at IS NULL AND pb.quantity_remaining > 0",
        {
          productId,
        },
      )
      .orderBy("pb.expiry_date", "ASC", "NULLS LAST")
      .addOrderBy("pb.id", "ASC")
      .getOne();

    product.nearestExpiryDate = earliestBatch ? earliestBatch.expiryDate : null;
    await manager.getRepository(Product).save(product);

    return savedBatch;
  }

  async restoreExactBatches(
    manager: EntityManager,
    orderItemId: number,
    productId: number,
  ): Promise<void> {
    const itemBatches = await manager.getRepository(OrderItemBatch).find({
      where: { orderItemId },
      order: { batchId: "ASC" },
    });

    if (itemBatches.length === 0) return;

    const product = await manager
      .getRepository(Product)
      .createQueryBuilder("p")
      .setLock("pessimistic_write")
      .where("p.id = :productId AND p.deleted_at IS NULL", { productId })
      .getOne();

    const batchRepo = manager.getRepository(ProductBatch);
    let totalRestored = 0;

    for (const ib of itemBatches) {
      const batch = await batchRepo
        .createQueryBuilder("pb")
        .setLock("pessimistic_write")
        .where("pb.id = :batchId", { batchId: ib.batchId })
        .getOne();

      if (batch) {
        batch.quantityRemaining += ib.quantityTaken;
        await batchRepo.save(batch);
        totalRestored += ib.quantityTaken;
      }
    }

    if (product && totalRestored > 0) {
      product.stockQuantity += totalRestored;

      const earliestBatch = await manager
        .getRepository(ProductBatch)
        .createQueryBuilder("pb")
        .where(
          "pb.product_id = :productId AND pb.deleted_at IS NULL AND pb.quantity_remaining > 0",
          { productId },
        )
        .orderBy("pb.expiry_date", "ASC", "NULLS LAST")
        .addOrderBy("pb.id", "ASC")
        .getOne();

      product.nearestExpiryDate = earliestBatch
        ? earliestBatch.expiryDate
        : null;
      await manager.getRepository(Product).save(product);
    }
  }

  async simulateFefo(
    productId: number,
    quantity: number,
  ): Promise<{ expiryDate: string | null; quantityTaken: number }[]> {
    if (quantity <= 0) return [];

    const batches = await this.productBatchRepository
      .createQueryBuilder("pb")
      .where(
        "pb.product_id = :productId AND pb.deleted_at IS NULL AND pb.quantity_remaining > 0",
        { productId },
      )
      .orderBy("pb.expiry_date", "ASC", "NULLS LAST")
      .addOrderBy("pb.id", "ASC")
      .getMany();

    let remaining = quantity;
    const result: { expiryDate: string | null; quantityTaken: number }[] = [];

    for (const batch of batches) {
      if (remaining <= 0) break;
      const take = Math.min(batch.quantityRemaining, remaining);
      result.push({ expiryDate: batch.expiryDate, quantityTaken: take });
      remaining -= take;
    }

    if (remaining > 0) {
      result.push({ expiryDate: null, quantityTaken: remaining });
    }

    return result;
  }

  async restoreQuantityForReturnedItem(
    manager: EntityManager,
    orderItemId: number,
    productId: number,
    quantityToRestore: number,
  ): Promise<void> {
    if (quantityToRestore <= 0) return;

    const product = await manager
      .getRepository(Product)
      .createQueryBuilder("p")
      .setLock("pessimistic_write")
      .where("p.id = :productId AND p.deleted_at IS NULL", { productId })
      .getOne();

    if (!product) return;

    const batchRepo = manager.getRepository(ProductBatch);
    const itemBatches = await manager.getRepository(OrderItemBatch).find({
      where: { orderItemId },
      order: { id: "DESC" },
    });

    let remainingToRestore = quantityToRestore;

    for (const ib of itemBatches) {
      if (remainingToRestore <= 0) break;

      const batch = await batchRepo
        .createQueryBuilder("pb")
        .setLock("pessimistic_write")
        .where("pb.id = :batchId", { batchId: ib.batchId })
        .getOne();

      if (batch) {
        batch.quantityRemaining += remainingToRestore;
        await batchRepo.save(batch);
        remainingToRestore = 0;
      }
    }

    if (remainingToRestore > 0) {
      const fallbackBatch = await batchRepo
        .createQueryBuilder("pb")
        .setLock("pessimistic_write")
        .where("pb.product_id = :productId AND pb.deleted_at IS NULL", {
          productId,
        })
        .orderBy("pb.expiry_date", "ASC", "NULLS LAST")
        .addOrderBy("pb.id", "DESC")
        .getOne();

      if (fallbackBatch) {
        fallbackBatch.quantityRemaining += remainingToRestore;
        await batchRepo.save(fallbackBatch);
      }
    }

    product.stockQuantity += quantityToRestore;

    const earliestBatch = await manager
      .getRepository(ProductBatch)
      .createQueryBuilder("pb")
      .where(
        "pb.product_id = :productId AND pb.deleted_at IS NULL AND pb.quantity_remaining > 0",
        { productId },
      )
      .orderBy("pb.expiry_date", "ASC", "NULLS LAST")
      .addOrderBy("pb.id", "ASC")
      .getOne();

    product.nearestExpiryDate = earliestBatch ? earliestBatch.expiryDate : null;
    await manager.getRepository(Product).save(product);
  }
}

File: backend-fastapi/app/modules/products/batch_consumption_service.py
from datetime import date, datetime, timezone
from decimal import Decimal
from typing import List, Optional, TypedDict

from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select

from app.core.exceptions import BusinessException
from app.modules.products.models import Product, ProductBatch
from app.modules.orders.models import OrderItemBatch


class SimulatedBatch(TypedDict):
    expiry_date: Optional[date]
    quantity_taken: int


class ConsumedBatch(TypedDict):
    batch_id: int
    quantity_taken: int
    expiry_date: Optional[date]


class BatchConsumptionService:
    def __init__(self, db: AsyncSession):
        self.db = db

    async def simulate_fefo(
        self, product_id: int, quantity: int
    ) -> List[SimulatedBatch]:
        if quantity <= 0:
            return []

        stmt = (
            select(ProductBatch)
            .where(
                ProductBatch.product_id == product_id,
                ProductBatch.deleted_at.is_(None),
                ProductBatch.quantity_remaining > 0,
            )
            .order_by(
                ProductBatch.expiry_date.asc().nulls_last(), ProductBatch.id.asc()
            )
        )
        batches = (await self.db.execute(stmt)).scalars().all()

        remaining = quantity
        result: List[SimulatedBatch] = []
        for batch in batches:
            if remaining <= 0:
                break
            take = min(batch.quantity_remaining, remaining)
            result.append({"expiry_date": batch.expiry_date, "quantity_taken": take})
            remaining -= take

        if remaining > 0:
            result.append({"expiry_date": None, "quantity_taken": remaining})

        return result

    async def list_batches(self, product_id: int) -> List[ProductBatch]:
        stmt = (
            select(ProductBatch)
            .where(
                ProductBatch.product_id == product_id,
                ProductBatch.deleted_at.is_(None),
            )
            .order_by(
                ProductBatch.expiry_date.asc().nulls_last(), ProductBatch.id.asc()
            )
        )
        return list((await self.db.execute(stmt)).scalars().all())

    async def receive_batch(
        self,
        product_id: int,
        quantity: int,
        expiry_date: Optional[date],
        unit_cost: Decimal,
        created_by: Optional[int],
        batch_code: Optional[str] = None,
    ) -> ProductBatch:
        product = await self._lock_product_or_throw(product_id)

        code = batch_code.strip() if batch_code and batch_code.strip() else None

        batch = ProductBatch(
            product_id=product_id,
            batch_code=code or "TEMP",
            quantity_received=quantity,
            quantity_remaining=quantity,
            unit_cost=unit_cost,
            expiry_date=expiry_date,
            created_by=created_by,
        )
        self.db.add(batch)
        await self.db.flush()

        if not code:
            now = datetime.now(timezone.utc)
            batch.batch_code = f"LÔ-{now:%Y%m%d}-{batch.id}"
            await self.db.flush()

        product.stock_quantity += quantity
        await self._refresh_nearest_expiry(product)

        return batch

    async def consume_fefo(self, product_id: int, quantity: int) -> List[ConsumedBatch]:
        if quantity <= 0:
            return []

        product = await self._lock_product_or_throw(product_id)

        stmt = (
            select(ProductBatch)
            .where(
                ProductBatch.product_id == product_id,
                ProductBatch.deleted_at.is_(None),
                ProductBatch.quantity_remaining > 0,
            )
            .order_by(
                ProductBatch.expiry_date.asc().nulls_last(), ProductBatch.id.asc()
            )
            .with_for_update()
        )
        batches = (await self.db.execute(stmt)).scalars().all()

        total_remaining = sum(b.quantity_remaining for b in batches)
        if total_remaining < quantity:
            raise BusinessException(
                "INVENTORY_INSUFFICIENT",
                409,
                f'Sản phẩm "{product.name}" không đủ tồn kho '
                f"(còn {total_remaining}, cần {quantity}).",
            )

        to_consume = quantity
        consumed: List[ConsumedBatch] = []
        for batch in batches:
            if to_consume <= 0:
                break
            take = min(batch.quantity_remaining, to_consume)
            batch.quantity_remaining -= take
            to_consume -= take
            consumed.append(
                {
                    "batch_id": batch.id,
                    "quantity_taken": take,
                    "expiry_date": batch.expiry_date,
                }
            )

        await self.db.flush()

        product.stock_quantity -= quantity
        await self._refresh_nearest_expiry(product)

        return consumed

    async def consume_specific_batch(
        self, product_id: int, batch_id: int, quantity: int
    ) -> List[ConsumedBatch]:
        if quantity <= 0:
            return []

        product = await self._lock_product_or_throw(product_id)

        stmt = (
            select(ProductBatch)
            .where(
                ProductBatch.id == batch_id,
                ProductBatch.product_id == product_id,
                ProductBatch.deleted_at.is_(None),
            )
            .with_for_update()
        )
        batch = (await self.db.execute(stmt)).scalars().first()

        if not batch:
            raise BusinessException(
                "BATCH_NOT_FOUND",
                404,
                f"Không tìm thấy lô hàng ID {batch_id} của sản phẩm này.",
            )

        if batch.quantity_remaining < quantity:
            raise BusinessException(
                "INVENTORY_INSUFFICIENT",
                409,
                f'Lô hàng "{batch.batch_code}" chỉ còn tồn '
                f"{batch.quantity_remaining} (cần trừ {quantity}).",
            )

        batch.quantity_remaining -= quantity
        await self.db.flush()

        product.stock_quantity -= quantity
        await self._refresh_nearest_expiry(product)

        return [
            {
                "batch_id": batch.id,
                "quantity_taken": quantity,
                "expiry_date": batch.expiry_date,
            }
        ]

    async def _lock_product_or_throw(self, product_id: int) -> Product:
        stmt = (
            select(Product)
            .where(Product.id == product_id, Product.deleted_at.is_(None))
            .with_for_update()
        )
        product = (await self.db.execute(stmt)).scalars().first()
        if not product:
            raise BusinessException(
                "PRODUCT_NOT_FOUND", 404, f"Không tìm thấy sản phẩm ID {product_id}."
            )
        return product

    async def _earliest_batch(self, product_id: int) -> Optional[ProductBatch]:
        stmt = (
            select(ProductBatch)
            .where(
                ProductBatch.product_id == product_id,
                ProductBatch.deleted_at.is_(None),
                ProductBatch.quantity_remaining > 0,
            )
            .order_by(
                ProductBatch.expiry_date.asc().nulls_last(), ProductBatch.id.asc()
            )
        )
        return (await self.db.execute(stmt)).scalars().first()

    async def _refresh_nearest_expiry(self, product: Product) -> None:
        earliest = await self._earliest_batch(product.id)
        product.nearest_expiry_date = earliest.expiry_date if earliest else None
        await self.db.flush()

    async def restore_exact_batches(self, order_item_id: int, product_id: int) -> None:
        stmt = (
            select(OrderItemBatch)
            .where(OrderItemBatch.order_item_id == order_item_id)
            .order_by(OrderItemBatch.batch_id.asc())
            .with_for_update()
        )
        item_batches = list((await self.db.execute(stmt)).scalars().all())
        if not item_batches:
            return

        product = await self._lock_product_or_throw(product_id)

        total_restored = 0
        for ib in item_batches:
            amount_to_restore = ib.quantity_taken - ib.restored_quantity
            if amount_to_restore <= 0:
                continue

            batch_stmt = (
                select(ProductBatch)
                .where(ProductBatch.id == ib.batch_id)
                .with_for_update()
            )
            batch = (await self.db.execute(batch_stmt)).scalars().first()
            if batch:
                batch.quantity_remaining += amount_to_restore
                total_restored += amount_to_restore

            ib.restored_quantity = ib.quantity_taken

        if total_restored > 0:
            product.stock_quantity += total_restored
            await self._refresh_nearest_expiry(product)

        await self.db.flush()

    async def restore_quantity_for_returned_item(
        self, order_item_id: int, product_id: int, quantity_to_restore: int
    ) -> None:
        if quantity_to_restore <= 0:
            return

        stmt = (
            select(Product)
            .where(Product.id == product_id, Product.deleted_at.is_(None))
            .with_for_update()
        )
        product = (await self.db.execute(stmt)).scalars().first()
        if not product:
            return

        item_batches_stmt = (
            select(OrderItemBatch)
            .where(OrderItemBatch.order_item_id == order_item_id)
            .order_by(OrderItemBatch.id.desc())
            .with_for_update()
        )
        item_batches = list((await self.db.execute(item_batches_stmt)).scalars().all())

        remaining_to_restore = quantity_to_restore

        for ib in item_batches:
            if remaining_to_restore <= 0:
                break

            available_in_this_batch = ib.quantity_taken - ib.restored_quantity
            if available_in_this_batch <= 0:
                continue

            restore_amount = min(available_in_this_batch, remaining_to_restore)

            batch_stmt = (
                select(ProductBatch)
                .where(ProductBatch.id == ib.batch_id)
                .with_for_update()
            )
            batch = (await self.db.execute(batch_stmt)).scalars().first()
            if batch:
                batch.quantity_remaining += restore_amount

            ib.restored_quantity += restore_amount
            remaining_to_restore -= restore_amount

        if remaining_to_restore > 0:
            fallback_stmt = (
                select(ProductBatch)
                .where(
                    ProductBatch.product_id == product_id,
                    ProductBatch.deleted_at.is_(None),
                )
                .order_by(
                    ProductBatch.expiry_date.asc().nulls_last(),
                    ProductBatch.id.desc(),
                )
                .with_for_update()
            )
            fallback_batch = (await self.db.execute(fallback_stmt)).scalars().first()
            if fallback_batch:
                fallback_batch.quantity_remaining += remaining_to_restore

        await self.db.flush()

        product.stock_quantity += quantity_to_restore
        await self._refresh_nearest_expiry(product)
